File: Baekjoon/yerim/부분수열의 합.cpp
// [1182] 부분수열의 합
#include <iostream>
using namespace std;

int n, s, ans, arr[21];

void input()
{
    cin >> n >> s;
    for(int i = 0; i < n; i++) {
        cin >> arr[i];
    }
}

void dfs(int idx, int sum, int cnt)
{
    if(idx == n) {
        if(sum == s && cnt >= 1) {
            ans++;
        }
        return;
    }

    dfs(idx + 1, sum, cnt);
    dfs(idx + 1, sum + arr[idx], cnt + 1);
}

int main(void)
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    input();
    dfs(0, 0, 0);
    cout << ans << '\n';

    return 0;
}

File: Baekjoon/chaeeun/부분수열의합.py
N, S = map(int, input().split())
nums = list(map(int, input().split()))

answer = 0
def dfs(idx, sum):
    global answer

    if idx >= N: # 수열의 마지막 인덱스까지 접근
        return
        
    sum += nums[idx]

    if S == sum: # 입력 받은 S 값과 sum이 같아질 경우
        answer += 1

    dfs(idx+1, sum-nums[idx]) # 현재 idx에 해당하는 원소 뺀 sum
    dfs(idx+1, sum) # 현재 idx에 해당하는 원소 포함한 sum

dfs(0,0)
print(answer)